contracts/PillarStaking.sol

```diff
@@ -0,0 +1,271 @@
+        if (_maxTotalStake > 0) maxTotalStake = _maxTotalStake;
+        stakingToken = IERC20(_stakingToken);
+        rewardToken = IERC20(_rewardToken);
+        setStateInitialized();
+    }
+
+    /* ========== MUTATIVE FUNCTIONS ========== */
+
+    function stake(uint256 _amount) external nonReentrant whenStakeable {
+        if (block.timestamp > stakingPeriod + STAKING_PERIOD)
+            revert("StakingPeriodPassed()");
+        if ((totalStaked + _amount) > maxTotalStake)
+            revert MaximumTotalStakeReached({
+                totalMaxStake: maxTotalStake,
+                currentStakedAmount: totalStaked,
+                remainingStakeableAmount: (maxTotalStake - totalStaked),
+                stakerAmount: _amount
+            });
+        if (_amount < minStake)
+            revert InvalidMinimumStake({minimumStakeAmount: minStake});
+        if (_amount > maxStake)
+            revert InvalidMaximumStake({maximumStakeAmount: maxStake});
+        if (_amount > maxStake - stakeholderData[msg.sender].stakedAmount)
+            revert StakeWouldBeGreaterThanMax();
+        if (stakingToken.balanceOf(msg.sender) < _amount)
+            revert InsufficientBalance();
+        totalStaked = totalStaked + _amount;
+        stakeholderData[msg.sender].stakedAmount =
+            stakeholderData[msg.sender].stakedAmount +
+            _amount;
+        if (!stakeholderData[msg.sender].staked) {
+            stakeholderList.push(msg.sender);
+        }
+        stakingToken.safeTransferFrom(msg.sender, address(this), _amount);
+        emit Staked(msg.sender, _amount);
+    }
+
+    function unstake()
+        external
+        nonReentrant
+        whenReadyForUnstake
+        whenRewardsAllocated
+    {
+        uint256 stakedBalance = stakeholderData[msg.sender].stakedAmount;
+        totalStaked = totalStaked - stakedBalance;
+        stakeholderData[msg.sender].stakedAmount = 0;
+        stakingToken.safeTransfer(msg.sender, stakedBalance);
+        uint256 rewardAmount = stakeholderData[msg.sender].rewardAmount;
+        stakeholderData[msg.sender].rewardAmount = 0;
+        rewardToken.safeTransfer(msg.sender, rewardAmount);
+        emit Unstaked(msg.sender, stakedBalance);
+        emit RewardPaid(msg.sender, rewardAmount);
+    }
+
+    /* ========== VIEWS ========== */
+
+    function getContractState() public view returns (StakingState) {
+        return state;
+    }
+
+    function getStakedAmountForAccount(address account)
+        public
+        view
+        returns (uint256 amount)
+    {
+        return stakeholderData[account].stakedAmount;
+    }
+
+    function getRewardAmountForAccount(address account)
+        external
+        view
+        returns (uint256 amount)
+    {
+        return stakeholderData[account].rewardAmount;
+    }
+
+    /* ========== RESTRICTED FUNCTIONS ========== */
+
+    function calculateRewardAllocation() external onlyOwner {
```

Comment:
We need a specific contract function to allocate reward tokens. This can be involved only by the owner. Allocation of rewards should trigger an Event.

Reply:
I am not sure what you mean. This fn allocates rewards to users and is invoked by only the owner. This calculates the reward amount and stores it against the account. Then when the user withdraws their stake it gives them their rewards. There is an event for reward allocation and an event for rewards granted (inside the unstake fn)

Reply:
The function is used to transfer WETH to the contract (however instead of just transferring the rewards, we call a specific contract function notifying that the rewards have been added to the staking contract and we are ready for reward distribution).